jujugui/static/gui/src/app/app.js

```diff
@@ -1588,7 +1588,80 @@ YUI.add('juju-gui', function(Y) {
       @method _handleNewModel
     */
     _handleNewModel: function() {
-      this.switchEnv(null, null, null, null, false, true);
+      this.switchEnv();
+    },
+
+    /**
+      Handle the request to display the user entity state.
+
+      @method _handleUserEntity
+      @param {Object} state - The application state.
+      @param {Function} next - Run the next route handler, if any.
+      @param {Boolean} lookupModels - Whether to fetch models.
+    */
+    _handleUserEntity: function(state, next, lookupModels=true) {
+      if (!this.userPaths) {
+        this.userPaths = {};
+      }
+      const userPath = state.user;
+      const match = this.userPaths[userPath];
+      if (match) {
+        if (match.type === 'store') {
+          this._renderCharmbrowser(state, next);
+        } else if (match.type === 'model') {
+          const uuid = match.model.uuid;
+          if (uuid === this.get('modelUUID')) {
+            // We're already connected/connecting to the model.
+            return;
+          }
+          this.set('modelUUID', uuid);
+          this.switchEnv(
+            this.createSocketURL(
+              this.get('socketTemplate'), uuid));
+        }
+      } else if (lookupModels) {
+        this.controllerAPI.listModelsWithInfo((err, modelList) => {
+          if (err) {
+            console.error('unable to list models', err);
+            this.db.notifications.add({
+              title: 'Unable to list models',
+              message: 'Unable to list models: ' + err,
+              level: 'error'
+            });
+            return;
+          }
+          modelList.forEach(model => {
+            this.userPaths[`${model.owner}/${model.name}`] = {
+              type: 'model',
+              model: model
+            };
+          });
+          // Don't look up models again otherwise we'll create an endless loop.
+          this._handleUserEntity(state, next, false);
```

Comment:
If you abstract the above mentioned code out into a function you could remove the extra `lookupModels` param and instead do:
```javascript
if (this.userPaths[userPath] === 'model') {
  connectToModel(...);
} else {
  // The code below in the else block
}
```
This will remove one nesting of conditional and remove the extra argument in the function signature.